### ISSUE TITLE
Add `tensor::cat` method

### ISSUE THREAD

**Issue by infalmo** [coreylowman/dfdx#463]

**Comment by infalmo** [comment]
I'd be more than willing to help (my project is blocked by this feature request!) if some guidance is given :)

**Comment by coreylowman** [comment]
Duplicates #43 

It'll likely look similar to the stack operation, which the main difference being the resulting Shape. The key piece will be adding something like:

```rust
trait Concat<Rhs: Shape>: Shape {
    type Combined: Shape;
    fn concat(&self, rhs: &Rhs) -> Self::Combined;
}
```

Then this can be implemented for different shapes like so:

```rust
impl Concat<(usize, )> for (usize, ) {
    type Combined = (usize, );
    fn concat(&self, rhs: &Rhs) -> Self::Combined {
     (self.0 + rhs.0,)
   }
}

#[cfg(feature = "nightly")]
impl<const M: usize, const N: usize> Concat<(Const<N>, )> for (Const<M>, ) {
    type Combined = (Const<{M + N}>, );
    ...
}
```
Note that const dims will require nightly.

For things with more than 1 dim, we should keep the rightmost dims all generic, e.g.:

```rust
impl<D2: Dim, D3: Dim, D4: Dim> Concat<(usize, D2, D3, D4)> for (usize, D2, D3, D4) { ... }
impl<const M: usize, const N: usize, D2: Dim, D3: Dim, D4: Dim> Concat<(Const<N>, D2, D3, D4)> for (Const<M>, D2, D3, D4) { ... }
```

Then you can add a TryConcat trait for device (similar to TryStack):

```rust
pub trait TryConcat<E: Dtype>: DeviceStorage {
    fn concat<A: Shape, B: Shape, T>(&self, a: Tensor<A, E, Self, T>, b: Tensor<B, E, Self, T>) -> Tensor<A::Combined, E, Self, T>
    where
        A: Concat<B>,
        T: Tape<Self> + Merge<T>;
}
```

I'm really hoping that a lot of the stack code can be used for this, but let me know as soon as you run into a roadblock

**Comment by infalmo** [comment]
So, I should just implement `concat` for the first dimension? Torch allows you to specify dimensions, tho I think a barebones one for now would be sufficient.

**Comment by infalmo** [comment]
Also, should I allow concatenation of an array of tensors like in `torch.cat` and `tensor::stack`?

**Comment by coreylowman** [comment]
I think first dim is a good to start with and should cover most use cases (e.g. you could always just permute the dim to the first if its not). We can expand the implementation later if needed.

It'd have to be a tuple of tensors unfortunately since all the shapes won't be the same, so not quite as simple as what is in stack (which supports arrays or vecs). Let's get the implementation of just concat-ing two tensors together working first, and then we can see if it'd be easy to convert it to using tuples.

**Comment by infalmo** [comment]
I'm a little confused. Does `impl<D2: Dim, D3: Dim, D4: Dim> Concat<(usize, D2, D3, D4)> for (usize, D2, D3, D4) { ... }` allow something like:
```rust
let a: Tensor<(usize, Const<2>)> = dev.zeros();
let b: Tensor<(Const<5>, Const<2>)> = dev.zeros();
Tensor::concat(a, b);
```

**Comment by infalmo** [comment]
Okay, the above would make it impossible to have `impl<const M: usize, const N: usize, D2: Dim, D3: Dim, D4: Dim> Concat<(Const<N>, D2, D3, D4)> for (Const<M>, D2, D3, D4) { ... }` which is a dealbreaker.

I did play around with `feature(specialization)`, but with no success (incomplete feature, throws compile errors). If nothing, it should be possible to implement the trait for all possibilities of first dims (aka, 4 implementations per shape)

**Comment by coreylowman** [comment]
If we wanted to mix const/usize (which in principal seems okay), you could add another trait to add two dimensions together:

```rust
trait AddDim<Rhs: Dim>: Dim {
    type Output: Dim;
    fn add(self, rhs) -> Self::Output;
}

impl AddDim<usize> for usize { ... }
impl AddDim<Const<N>> for usize { type Output = usize; ... }
// etc
```
Then you could just do the generic dimensions like you said

**Comment by infalmo** [comment]
> If we wanted to mix const/usize (which in principal seems okay), you could add another trait to add two dimensions together:
> 
> ```rust
> impl std::ops::Add<usize> for usize { ... }
> impl std::ops::Add<Const<N>> for usize { type Output = usize; ... }
> impl std::ops::Add<usize> for Const<N> { type Output = usize; ...
> // etc
> ```
> 
> Then you could just do the generic dimensions like you said

I think this warrants aliasing `usize` to `Dyn`. Adding `Const<N>` to usize looks very weird
